File: DataStructures/LinkedList/swapNodes.cpp
#include <cstdlib>
#include <iostream>

using namespace std;
class Node{
    public:
    int data;
    Node * next;
    Node(int d)
    {
        data = d;
    }
};


void swapNodes(Node ** head_ref, int x, int y)
{
    if(x == y)
    {
        return;
        // nothing to do if x and y are the same
    }

    //Search for x (keep track of prev and curr x)
    Node * prevX = nullptr;
    Node * currX = *head_ref;

    while(currX && currX->data !=x)
    {
        prevX = currX;
        currX = currX->next;
    }

    //search fo y 
    Node * prevY = nullptr;
    Node * currY = *head_ref;
    while(currY && currY->data != y)
    {
        prevY = currY;
        currY = currY->next;
    }

    //if either y or x isn't present

    if(currX == nullptr || currY == nullptr)
    {
        return;
    }
//IMPORTANT
//this swaps x and y 
    // if x isn't head of the linked list
    if(prevX != nullptr)
    {
        prevX->next = currY;
    }else{ //else make y as new head
        *head_ref = currY;
    }

    if(prevY != nullptr)
    {
        prevY -> next = currX;
    }else{
        *head_ref = currX;
    }



    //Swap next pointers
    //swap pointers//or variables with a temp variable

    Node * temp = currY->next;
    currY->next = currX->next;
    currX->next = temp;
    //swapping currents 
}

void push(Node ** head_ref, int new_data)
{
    Node * new_node = new Node(new_data);
    new_node->next = *head_ref;
    *head_ref = new_node;
}

void printList(Node* head)
{
    while(head != nullptr)
    {
        cout << head->data << " ";
        head = head->next;
    }
}

int main()
{
    Node * head = new Node(7);

    push(&head, 2);
    push(&head, 21);
    push(&head, 26);
    push(&head, 23);
    push(&head, 6);

    cout << "Linkedlist before swapping nodes" << endl;

    printList(head);

    swapNodes(&head, 26, 21);
    cout << "LinkedList after calling swapNodes()"<<endl;
    printList(head);

}

File: DataStructures/LinkedList/swapNodes.cs
using System;

class LinkedList{

    class Node{
        int data;
        Node next;
        Node(int d)
        {
            data = d;
            next = null;
        }
    }

    LinkedList(int h)
    {
        Node head = Node(h);
    }

    public void swapNodes(int x, int y)
    {
        if(x == y)
        {
            return;
        }

        Node currX = head;
        Node prevX = null;

        while(currX != null && currX.data != x)
        {
            prevX = currX;
            currX = currX.next;
        }

        Node prevY = null;
        Node currX = head;

        while(currY != null && currY.data != y)
        {
            prevY = currY;
            currY = currY.next;
        }

        if(currX == null || currY == null)
        {
            return;
        }

        if(prevX != null)
        {
            prevX.next = currY;
        }else{
            currY = head;
        }

        if(prevY != null)
        {
            prevY.next = currX;
        }else{
            currX = head;
        }

        Node temp = currX.next;
        currX.next = currY.next;
        currY.next = temp;
    }

    public void push(int newData)
    {
        Node new_node = new Node(newData);
        new_node.next = head;
        head = new_node;
    }

    public void printList()
    {
        node = head;
        while(node != null)
        {
            Console.Write(node.data + " ");
            node = node.next;
        }
    }
}

public static void Main(String[] args)
{
    LinkedList llist = new LinkedList()
    llist.push(17)
    llist.push(37)
    llist.push(73)
    llist.push(27)
    llist.push(71)
    llist.push(87)
    Console.Write("Linked List before calling Swap Nodes")
    llist.printList();
    llist.swapNodes(17, 37);
    Console.Write("Linked List after calling swap nodes");
    llist.printList();
}